pkg/proxy/nftables/proxier.go

```diff
@@ -0,0 +1,1631 @@
+}
+
+// nftablesBaseChains lists our "base chains"; those that are directly connected to the
+// netfilter hooks (e.g., "postrouting", "input", etc.), as opposed to "regular" chains,
+// which are only run when a rule jumps to them. See
+// https://wiki.nftables.org/wiki-nftables/index.php/Configuring_chains.
+//
+// These are set up from setupNFTables() and then not directly referenced by
+// syncProxyRules().
+//
+// All of our base chains have names that are just "${type}-${hook}". e.g., "nat-prerouting".
+type nftablesBaseChain struct {
+	name      string
+	chainType knftables.BaseChainType
+	hook      knftables.BaseChainHook
+	priority  knftables.BaseChainPriority
+}
+
+var nftablesBaseChains = []nftablesBaseChain{
+	// We want our filtering rules to operate on pre-DNAT dest IPs, so our filter
+	// chains have to run before DNAT.
+	{"filter-input", knftables.FilterType, knftables.InputHook, knftables.DNATPriority + "-1"},
+	{"filter-forward", knftables.FilterType, knftables.ForwardHook, knftables.DNATPriority + "-1"},
+	{"filter-output", knftables.FilterType, knftables.OutputHook, knftables.DNATPriority + "-1"},
+	{"nat-prerouting", knftables.NATType, knftables.PreroutingHook, knftables.DNATPriority},
+	{"nat-output", knftables.NATType, knftables.OutputHook, knftables.DNATPriority},
+	{"nat-postrouting", knftables.NATType, knftables.PostroutingHook, knftables.SNATPriority},
+}
+
+// nftablesJumpChains lists our top-level "regular chains" that are jumped to directly
+// from one of the base chains. These are set up from setupNFTables(), and some of them
+// are also referenced in syncProxyRules().
+type nftablesJumpChain struct {
+	dstChain  string
+	srcChain  string
+	extraArgs string
+}
+
+var nftablesJumpChains = []nftablesJumpChain{
+	{kubeEndpointsCheckChain, "filter-input", "ct state new"},
+	{kubeEndpointsCheckChain, "filter-forward", "ct state new"},
+	{kubeEndpointsCheckChain, "filter-output", "ct state new"},
+
+	{kubeForwardChain, "filter-forward", ""},
+
+	{kubeFirewallCheckChain, "filter-input", "ct state new"},
+	{kubeFirewallCheckChain, "filter-output", "ct state new"},
+	{kubeFirewallCheckChain, "filter-forward", "ct state new"},
+
+	{kubeServicesChain, "nat-output", ""},
+	{kubeServicesChain, "nat-prerouting", ""},
+	{kubeMasqueradingChain, "nat-postrouting", ""},
+}
+
+// ensureChain adds commands to tx to ensure that chain exists and doesn't contain
+// anything from before this transaction.
+func ensureChain(chain string, tx *knftables.Transaction, existing sets.Set[string]) {
+	if existing.Has(chain) {
+		return
+	}
+	tx.Add(&knftables.Chain{
+		Name: chain,
+	})
+	tx.Flush(&knftables.Chain{
+		Name: chain,
+	})
+	existing.Insert(chain)
+}
+
+func (proxier *Proxier) setupNFTables(tx *knftables.Transaction) {
+	ipX := "ip"
+	ipvX_addr := "ipv4_addr" //nolint:stylecheck // var name intentionally resembles value
+	noLocalhost := "ip daddr != 127.0.0.0/8"
+	if proxier.ipFamily == v1.IPv6Protocol {
```

Comment:
that is unrelated to this PR but to other comment in my PR, v1.IPFamily is an enum with 3 values, and we have "things" that can only be ipv4 or ipv6, I understand that semantically this looks more correct but a boolean is more acccurate, if is not IPv4 then it is IPv6, it can not be unknown

Reply:
`IPFamilyUnknown` is essentially an error value. It should not exist anywhere other than as the return value of a function that is also returning an `error`.

`proxier.ipFamily` _cannot_ be `IPFamilyUnknown`, any more than `proxier.nftables` can be `nil`.

Reply:
(Oh, but also, I don't really love the `ipX`/`ipvX_addr` stuff and am looking for suggestions for improving that... that was a late change after I realized that using `-D IP=ip`/`-D IP=ip6` and `$IP` wouldn't actually work like I thought it would.)

Reply:
it seems is a personal taste then